design/ironic-interaction/ironic-background-docs.md

```diff
@@ -0,0 +1,346 @@
+   to disk. This is performed as a HTTP PATCH request to the
+   /v1/nodes/node-name endpoint.
+
+       {
+       {“op”: “replace”, “path”: “/instance_info”, “value”: {
+           “image_source”: “http://url-to-image”,
+           “image_os_hash_algo”: “sha256”,
+           “image_os_hash_value”: “abcdefghi…”}},
+       {“op”: “replace”, “path”: “/instance_uuid”, “value”: “anyuuidvalue”}},
+       }
+
+   **NOTE:** Instead of defining the "image_os_hash_*" values, a MD5 based
+   image checksum can be set.
+
+   This configuration does two things. First sets the image and checksum
+   to be utilized for image verification, and sets an "instance_uuid" value
+   which acts as a signal to any client that the node has been claimed by
+   an API cient. The instancce_uuid can be set to any value, and is
+   ultimately not required.
+
+2. Request ironic to perform a "validate" operation on the information
+   it is presently configured with. The expected response is a HTTP 200
+   return code, with a message body that consists of a list of "driver"
+   interfaces and any errors if applicable.
+
+    GET /v1/nodes/node-name/validate
+
+   Reply: 
+
+       {
+       "boot": true,
+       ..
+       "deploy": "configuration error message if applicable"
+       }
+
+   The particular interfaces that would be important to pay attention to are
+   ‘boot’, ‘deploy’, ‘power’, ‘management’.
+
+   More information can be found at: https://developer.openstack.org/api-ref/baremetal/?expanded=validate-node-detail
+
+3. Craft a configuration drive file
+
+   Configuration drives are files that contain a small ISO9660 filesystem
+   which contains configuration metadata and user defined "user-data".
+
+   1) Create a folder called “TEMPDIR”
+   2) In the case of ignition based configuration, that file would be
+      placed renamed "user_data" and placed in TMPDIR/openstack/latest/
+      folder.
+   3) Metadata for networking configuration setup using "cloud-init" or
+      cloud-init similar application would also be written to the
+      TEMPDIR/openstack/latest as well. This is out of scope, but is well
+      documented in the OpenStack community.
+   4) TEMPDIR is converted to an iso9660 image using a label of “config-2”.
+   5) The resulting ISO9660 image file is then compressed using the gzip
+      algorithm.
+   6) The resulting gzip image file is then encoded in base64 for storage
+      and transport. Ironic does the needful to decode and uncompress the
+      configuration drive prior to deployment.
+
+
+4. Send a HTTP POST to /v1/nodes/node-name/states/provision to initiate
+   the deployment
+
+       {“target”: “active”,
+        “configdrive”: “http://url-to-config-drive/node.iso.gz”}
+
+   Once the request to make the node active has been received by ironic,
+   it will proceed with the deployment process and execute the required
+   steps to help ensure the baremetal node reboots into the requested
+   disk image.
```

Comment:
How do we monitor progress?

Reply:
A caller can reference the state machine and the ?deploy_step? field on the node object. That being said, there are only a couple deployment classified steps until we continue to evolve the ability to define a greater amount of detail for how the node is to be deployed (mainly consisting of raid configuration details).